Add localization to core-interop and hierarchy-builder packages

JonasDov opened the pull request:
prerequisite #242

grigasp commented on packages/hierarchy-builder/public/locales/en/PresentationHierarchyBuilder.json:
I believe this file is not going to be included in the package. Please see `.npmignore` in `presentation-components` for reference.

JonasDov replied:
Added .npmignore

grigasp commented on packages/core-interop/src/core-interop/Localization.ts:
```diff
@@ -0,0 +1,14 @@
+ * @beta
+ */
+export function createTranslator(localization: Localization): (input: string) => string {
+  return (input) => localization.getLocalizedString(input);
```
I believe here we could take care of registering the `hierarchy-builder` localization namespace (that would mean `createTranslator` has to become `async`).

JonasDov replied:
Added

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/Localization.ts:
```diff
@@ -0,0 +1,39 @@
+ *--------------------------------------------------------------------------------------------*/
+
+/** A translator implementation returns the same input. */
+const NOOP_Translator = (input: string) => input;
```
I think this variable should be all in upper case.

JonasDov replied:
Renamed

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/Localization.ts:
```diff
@@ -0,0 +1,39 @@
+ * Set translator to use by this package. By default the package uses a no-op translator.
+ * @beta
+ */
+export function setTranslator(translator: Translator) {
+  g_translator = translator;
+}
```
we should allow setting to `undefined`, which would reset back to noop localization function

JonasDov replied:
Made localizationFunction parameter optional

grigasp reviewed (changes requested):
- [ ] Maybe rename all instances of "Translator" to "LocalizationFunction"?
- [ ] Could we also have 1 full stack test that:
  1. initializes `IModelApp`
  2. sets the `hierarchy-builder` localization function
  3. gets some localized string and ensures it's localized.

JonasDov commented:
> * [ ]  Maybe rename all instances of "Translator" to "LocalizationFunction"?
> * [ ]  Could we also have 1 full stack test that:
>   
>   1. Initializes `IModelApp`;
>   2. Sets the `hierarchy-builder` localization function;
>   3. Gets some localized string and ensures it's localized.

* [x] Renamed
* [x] Added full stack test

grigasp commented on packages/core-interop/src/test/Localization.test.ts:
```diff
@@ -0,0 +1,24 @@
+      getLocalizedString: (input: string) => `${input}_localized`,
+      registerNamespace: async () => {},
+    } as unknown as Localization;
+    const spy = sinon.spy(localization, "registerNamespace");
```
could just set the spy right away:
```ts
    const localization = {
      getLocalizedString: (input: string) => `${input}_localized`,
      registerNamespace: sinon.spy(),
    } as unknown as Localization;
```

JonasDov replied:
Changed to:
```ts
const registerNamespaceSpy = sinon.spy();
const localization = {
  getLocalizedString: (input: string) => `${input}_localized`,
  registerNamespace: registerNamespaceSpy,
} as unknown as Localization;
const translator = await createLocalizationFunction(localization);
expect(registerNamespaceSpy).to.be.calledWith(LOCALIZATION_NAMESPACE);
```
Because I am checking what the spy was called with.

grigasp commented on packages/full-stack-tests/src/hierarchy-builder/Localization.test.ts:
```diff
@@ -0,0 +1,32 @@
+
+describe("Stateless hierarchy builder", () => {
+  describe("Localization", () => {
+    describe("translate", () => {
```
I don't think we need this "translate" describe.

JonasDov replied:
Removed

grigasp commented on packages/full-stack-tests/src/hierarchy-builder/Localization.test.ts:
```diff
@@ -0,0 +1,32 @@
+        await terminate();
+      });
+
+      it("translates input based on localization", async function () {
```
```suggestion
      it("translates strings using `IModelApp.localization`", async function () {
```

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/Localization.ts:
```diff
@@ -0,0 +1,38 @@
+const NOOP_LOCALIZATION_FUNCTION = (input: string) => input;
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+let g_localization_function = NOOP_LOCALIZATION_FUNCTION;
```
```suggestion
let g_localizationFunction = NOOP_LOCALIZATION_FUNCTION;
```
